### PR TITLE
Simplify some iterator combinators

### PR DESCRIPTION
**Pull request by compiler-errors**
Saw a `flat_map().next()` instead of a `find_map()` in some method probe code, so did a quick search for some similar usages.

### CODE REVIEW

**Comment by compiler-errors** [comment]
@leonardo-m no -- clippy doesn't trigger on this:

```rust
fn main() {
    let x = vec![()];
    x.into_iter().flat_map(|_| Some(1)).next();
    // ^ should probably suggest `find_map`.
}
```

**Comment by compiler-errors** [comment]
So `flat_map()` over `Option<_>` is a pedantic lint already that suggests using `filter_map`, and then `filter_map(..).next()` is also a lint. So I'm not sure there's much else to suggest for clippy!

**Comment by vacuus** [comment]
In `compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs`, you can change `.any(..)` to `.all(..)` and remove both negations. Is Clippy missing this?

**Comment by SkiFire13** [comment]
> In `compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs`, you can change `.any(..)` to `.all(..)` and remove both negations. Is Clippy missing this?

Note that there's also an `&&` in the middle between two negations, so you need to remove all three negations and change the `&&` to a `||`. I guess that's too much for clippy.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -515,26 +515,21 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+            // let x;
+            // x += 1;
+            // ```
+            !i.contains(span)
                     // We filter these to avoid incorrect main message on `match-cfg-fake-edges.rs`
                         && !visitor
                             .errors
                             .iter()
                             .map(|(sp, _)| *sp)
                             .any(|sp| span < sp && !sp.contains(span))
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:531]: The whitespace seems a bit odd.

File: compiler/rustc_borrowck/src/diagnostics/region_errors.rs [hunk 2]
```diff
@@ -298,35 +298,29 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
-                                        let mut traits = vec![];
-                                        let mut hir_v = HirTraitObjectVisitor(&mut traits, *did);
-                                        hir_v.visit_ty(self_ty);
-                                        !traits.is_empty()
-                                    }) =>
-                                    {
-                                        Some(self_ty)
-                                    }
-                                    _ => None,
+                        match tcx.hir().trait_impls(trait_did).iter().find_map(|&impl_did| {
```
Comment (cjgillot) [inline compiler/rustc_borrowck/src/diagnostics/region_errors.rs:301]: A scrutinee of this size probably deserves a `let` binding with a descriptive name.
This whole function probably deserves some application of the `?` operator.

File: compiler/rustc_infer/src/infer/error_reporting/nice_region_error/static_impl_trait.rs [hunk 3]
```diff
@@ -428,35 +415,29 @@ impl<'a, 'tcx> NiceRegionError<'a, 'tcx> {
...
-                                        let mut traits = vec![];
-                                        let mut hir_v = HirTraitObjectVisitor(&mut traits, *did);
-                                        hir_v.visit_ty(self_ty);
-                                        !traits.is_empty()
-                                    }) =>
-                                    {
-                                        Some(self_ty)
-                                    }
-                                    _ => None,
+                        match tcx.hir().trait_impls(trait_did).iter().find_map(|&impl_did| {
```
Comment (cjgillot) [inline compiler/rustc_infer/src/infer/error_reporting/nice_region_error/static_impl_trait.rs:418]: Like earlier, a `let` binding and `?`s may be useful.

File: compiler/rustc_lint/src/unused.rs [hunk 4]
```diff
@@ -270,22 +270,17 @@ impl<'tcx> LateLintPass<'tcx> for UnusedResults {
...
-                    .next(),
+                ty::Dynamic(binders, _, _) => binders.iter().find_map(|predicate| {
+                    if let ty::ExistentialPredicate::Trait(ref trait_ref) = predicate.skip_binder()
+                    {
+                        let def_id = trait_ref.def_id;
+                        is_def_must_use(cx, def_id, span)
+                    } else {
+                        None
+                    }
+                    .map(|inner| MustUsePath::TraitObject(Box::new(inner)))
```
Comment (cjgillot) [inline compiler/rustc_lint/src/unused.rs:282]: Could you move this `.map` in the first branch, since it's the only case where it does anything?

**Review by cjgillot — COMMENTED** [review]
This is a well-deserved simplifications. A few complementary ideas.

**Review by cjgillot — APPROVED** [review]
A few extra API nits and r=me.

**Comment by compiler-errors** [comment]
Addressed comments in `static_impl_trait.rs`.

@bors r=cjgillot

**Pull request merged**